### DISCUSSION TITLE
블로킹과 논블로킹에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by nayonsoso** [2024-woowacourse-study/level-interview#138] [Network]
.

**Comment by le2sky** [comment]
블로킹, 논블로킹, 동기, 비동기
위에 적힌 4개에 대해서 다루면 더 좋을 것 같네요.

**Comment by jimi567** [comment]
### Blocking / Non-Blocking 
블로킹은 다른 요청의 작업을 처리하기 위해 현재 작업을 block(waiting) 하는 것  
논블로킹은 block(waiting)하지 않는 것  
즉, 블로킹/논블로킹은 전체적인 작업의 **제어권(흐름)을 막냐 막지 않느냐의 차이**
 
![image](https://github.com/2024-woowacourse-study/level-interview/assets/28584160/28e1c88c-58b1-467d-9259-778ca2678b2f) 

### 동기/ 비동기
동기는 작업 시간을 함께 맞춰서 실행한다는 뜻,   
**즉, 요청한 작업에 대해 완료 여부를 따져 순차적으로 처리하는 것을 의미함**
비동기는  **요청한 작업에 대해 완료 여부를 따지지 않음** 따라서 자신의 다음 작업을 그대로 수행함  
여러개의 요청 작업을 순차적으로 처리 하냐 아니냐의 차이  
**동기는 요청한 작업에 대해 순서가 지켜지고, 비동기는 순서가 지켜지지 않음**

![image](https://github.com/2024-woowacourse-study/level-interview/assets/28584160/4df7c17a-dd51-43e9-9240-4ae4993b479f)  

비동기의 경우 I/O 작업과 같은 오버헤드가 큰 작업을 수행할 때, 해당 I/O와 관련 없이 다음 작업을 수행할 수 있다면, 응답 시간을 줄이는데 큰 이점이 있음  

예를 들어, 요청에 대한 서버의 작업 중 DB 쿼리를 수행하는 작업이 있을 경우  
이 작업을 동기로 수행하면 데이터베이스에서 응답이 올 때 까지 대기해야 함  
이러면 다른 요청은 처리하지 못하므로, 대규모 트래픽이 발생하게 되면, 애플리케이션의 성능이 크게 저하될 수 있음  
하지만 비동기로 작업을 수행하게 되면, 데이터베이스 응답이 올 때 까지 대기하지 않고 다른 요청을 동시에 처리할 수 있게 되므로 응답 시간에 이점이 있다.  


### 동기/비동기 + 블로킹/논블로킹 조합
![image](https://github.com/2024-woowacourse-study/level-interview/assets/28584160/36628135-3730-4cfb-ba30-f261d55fa430)


> 동기 + 블로킹 조합

다른 작업이 진행되는 동안 자신의 작업을 처리하지 않고 (Blocking), 다른 작업의 완료 여부를 바로 받아 순차적으로 처리하는 (Sync) 방식  **다른 작업의 결과가 자신의 작업에 영향을 주는 경우에 활용할 수 있다.**


예를 들면,  DB에서 데이터를 조회해야 다음 작업을 처리할 수 있는 경우를 생각하면 됨  

> 비동기 + 논블로킹 조합  

다른 작업이 진행되는 동안에도 자신의 작업을 처리하고 (Non Blocking), 다른 작업의 결과를 바로 처리하지 않아 작업 순서가 지켜지지 않는 (Async) 방식 **다른 작업의 결과가 자신의 작업에 영향을 주지 않은 경우에 활용할 수 있다.**  

작업량이 많거나, 시간이 오래 걸리는 작업을 처리해야 하는 경우 고려할 수 있음  
예를 들어, 대용량 데이터를 처리하거나 많은 트래픽을 처리하는 서비스에서 한 작업이 처리되는 동안 다른 작업을 처리할 수 있으므로 전체적인 응답 시간을 줄일 수 있어 애플리케이션의 성능을 향상시킬 수 있음

> 동기 + 논블로킹 조합 

다른 작업이 진행되는 동안에도 자신의 작업을 처리하고 (Non Blocking), 다른 작업의 결과를 바로 처리하여 작업을 순차대로 수행 하는 (Sync) 방식


> 비동기 + 블로킹 조합   

다른 작업이 진행되는 동안 자신의 작업을 멈추고 기다리는 (Blocking), 다른 작업의 결과를 바로 처리하지 않아 순서대로 작업을 수행하지 않는 (Async) 방식

**Comment by GIVEN53** [comment]
### Sync / Async
호출하는 함수의 작업 완료를 기다리는지 여부의 차이가 있습니다.
- 함수 A가 **동기**로 함수 B를 호출하면 A는 B의 작업 완료이 완료될 때까지 기다려야 합니다. 따라서 작업이 순차적으로 진행됩니다.
- 함수 A가 **비동기**로 함수 B를 호출하면 A는 B의 작업 완료를 신경쓰지 않고 따로 동작합니다. 따라서 작업이 순차적으로 진행되지 않습니다.

따라서 동기, 비동기는 **작업 순서**와 관련된 개념입니다.

### Blocking / Non-blocking
호출하는 함수에게 제어권을 넘기는지 여부의 차이가 있습니다.
> 제어권 : 함수의 코드나 프로세스의 실행 흐름을 제어할 수 있는 권리

- 함수 A가 함수 B를 블로킹으로 호출하면 제어권을 B에게 넘기고 실행을 멈춥니다. 함수 B의 실행이 끝나면 제어권을 돌려받아 멈춘 작업을 진행합니다.
- 함수 A가 함수 B를 논블로킹으로 호출하면 제어권을 B에게 넘기지 않고 계속 진행합니다. 

따라서 블로킹, 논블로킹은 **병렬 실행**과 관련된 개념입니다.

### Sync + Blocking
- 손님 : 아메리카노 한 잔 주세요 -> `Sync`
- (직원은 아메리카노를 제조하고 손님은 기다린다) -> `Blocking`
- 직원 : 아메리카노 한 잔 나왔습니다 -> `작업 종료`
- 손님 : 소금빵 한 개 주세요 -> `Sync`
- (직원은 소금빵을 포장하고 손님은 기다린다) -> `Blocking`
- 직원 : 소금빵 한 개 나왔습니다 -> `작업 종료`
- (손님은 자리로 간다) -> `다음 작업`

### Sync + Non-blocking
- 손님 : 아메리카노 한 잔 주세요 -> `Sync`
- (직원은 아메리카노를 제조한다)
- 손님 : 다 됐나요? -> `Non-blocking`
- 직원 : 아메리카노 한 잔 나왔습니다 -> `작업 종료`
- 손님 : 소금빵 한 개 주세요 -> `Sync`
- (직원은 소금빵을 포장한다)
- 손님 : 다 됐나요? -> `Non-blocking`
- 바리스타 : 아직 포장 중입니다
- 손님 : 다 됐나요? -> `Non-blocking`
- 바리스타 : 아직 포장 중입니다
- 손님 : 다 됐나요? -> `Non-blocking`
- 직원 : 소금빵 한 개 나왔습니다 -> `작업 종료`
- (손님은 자리로 간다) -> `다음 작업`

### Async + Blocking
- 손님 : 아메리카노 한 잔 주세요 -> `Async`
- (직원은 아메리카노를 제조하고 손님은 기다린다) -> `Blocking`
- 직원 : 아메리카노 한 잔 나왔습니다 -> `작업 종료`
- 손님 : 소금빵 한 개 주세요 -> `Async`
- (직원은 소금빵을 포장하고 손님은 기다린다) -> `Blocking`
- 직원 : 소금빵 한 개 나왔습니다 -> `작업 종료`
- (손님은 자리로 간다) -> `다음 작업`

> Sync + Blocking 방식과 개념적 차이만 있고, 성능상의 차이는 거의 없다.
> 의도적으로 구현하기 보다는 Async + Non-blocking으로 처리하려다 개발자의 실수로 Blocking으로 처리해서 이러한 결과가 발생한다.

### Async + Non-blocking
- 손님 : 아메리카노 한 잔 주세요 -> `Async`
- 손님 : 소금빵 한 개 주세요 -> `Async`
- (손님은 자리로 간다) -> `Non-blocking`
- (직원은 아메리카노를 제조한다)  
- (직원은 소금빵을 포장한다)
- 직원 : 소금빵 한 개 나왔습니다 -> `작업 종료`
- 직원 : 아메리카노 한 잔 나왔습니다 -> `작업 종료`

**Comment by alstn113** [comment]
둘 다 예시가 있어서 읽기가 편하네요. ^_____^
